8324677: Specification clarification needed for JVM TI GetObjectMonitorUsage

sspitsyn opened the pull request:
The implementation of the JVM TI `GetObjectMonitorUsage` does not match the spec.
The function returns the following structure:
```
typedef struct {
    jthread owner;
    jint entry_count;
    jint waiter_count;
    jthread* waiters;
    jint notify_waiter_count;
    jthread* notify_waiters;
} jvmtiMonitorUsage;
```

The following four fields are defined this way:
```
waiter_count             [jint]         The number of threads waiting to own this monitor
waiters                      [jthread*] The waiter_count waiting threads
notify_waiter_count  [jint]         The number of threads waiting to be notified by this monitor
notify_waiters	[jthread*]           The notify_waiter_count threads waiting to be notified
```
The `waiters` has to include all threads waiting to enter the monitor or to re-enter it in `Object.wait()`.
The implementation also includes the threads waiting to be notified in `Object.wait()` which is wrong.
The `notify_waiters` has to include all threads waiting to be notified in `Object.wait()`.
The implementation also includes the threads waiting to re-enter the monitor in `Object.wait()` which is wrong.
This update makes it right.

The implementation of the JDWP command `ObjectReference.MonitorInfo (5)` is based on the JVM TI `GetObjectMonitorInfo()`. This update has a tweak to keep the existing behavior of this command.

The follwoing JVMTI vmTestbase tests are fixed to adopt to the `GetObjectMonitorUsage()` correct behavior:
```
  jvmti/GetObjectMonitorUsage/objmonusage001
  jvmti/GetObjectMonitorUsage/objmonusage003
```

The following JVMTI JCK tests have to be fixed to adopt to correct behavior:
```
vm/jvmti/GetObjectMonitorUsage/gomu001/gomu00101/gomu00101.html
vm/jvmti/GetObjectMonitorUsage/gomu001/gomu00101/gomu00101a.html
vm/jvmti/GetObjectMonitorUsage/gomu001/gomu00102/gomu00102.html
vm/jvmti/GetObjectMonitorUsage/gomu001/gomu00102/gomu00102a.html

```

A JCK bug will be filed and the tests have to be added into the JCK problem list located in the closed repository by a separate sub-task before integration of this update.

Also, please see and review the related CSR:
 [8324677](https://bugs.openjdk.org/browse/JDK-8324677) Specification clarification needed for JVM TI GetObjectMonitorUsage
 
Testing:
 - tested with mach5 tiers 1-6

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1549,16 +1549,21 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
             Handle th(current_thread, get_vthread_or_thread_oop(w));
-            ret.waiters[offset + j] = (jthread)jni_reference(calling_thread, th);
-            ret.notify_waiters[j++] = (jthread)jni_reference(calling_thread, th);
+            if (java_lang_Thread::get_thread_status(w->threadObj()) ==
+                JavaThreadStatus::BLOCKED_ON_MONITOR_ENTER) {
```
I don't think this is possible. `BLOCKED_ON_MONITOR_ENTER` is only set after the target has been removed from the wait-set and added to the cxq queue - see `ObjectMonitor::INotify`. As per the comment just above:
```
// If the thread was found on the ObjectWaiter list, then
// it has not been notified.
```
which means it can't be trying to acquire the monitor either.

sspitsyn replied:
Thank you for looking at it and the comment, David! Will check what is possible here.
Do you have any suggestions?

dcubed-ojdk reviewed (changes requested):
Okay... so we might have an incorrect implementation of JVM TI
GetObjectMonitorUsage, but I'm not convinced that it is broken
in the way that we think it is broken.

Please read thru my unfortunately long comments on the complicated
logic in `JvmtiEnvBase::get_object_monitor_usage()`. I _think_ things
are broken in a different way than what we're talking about in this bug.

In particular, the logic that populates the `waiters` array and reduces
the `waiter_count` value overrides the value that we think we're fixing
in the beginning of the function.

Similarly, the logic that populates the `notify_waiters` array and reduces
the `notify_waiter_count` value also overrides the value that we think
we're fixing in the beginning of the function.

dholmes-ora commented:
Based on Dan's analysis I would have to go back and redo the complete analysis for myself. :( I think the only way to make sense of this is to actually set up scenarios where we have different threads contending on entry, different threads waiting and different threads re-entering after being notified, and see what values actually get returned in each case.

I think the `pending_current_monitor` issue may be a distinct/different issue. I can easily imagine that this was overlooked when we introduced the "wait morphing" rather than have the notified/timed-out/interrupted waiters actually place themselves on the entry queue.

sspitsyn commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003/objmonusage003.cpp:
```diff
@@ -146,9 +146,9 @@ Java_nsk_jvmti_GetObjectMonitorUsage_objmonusage003_check(JNIEnv *env,
 
-    if (inf.waiter_count != waiterCount) {
+    if (inf.notify_waiter_count != notifyWaiterCount) {
         printf("(%d) waiter_count expected: %d, actually: %d\n",
```
Nice catch, thanks!

sspitsyn commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage001/objmonusage001.cpp:
```diff
@@ -94,7 +94,7 @@ jint  Agent_Initialize(JavaVM *jvm, char *options, void *reserved) {
 JNIEXPORT void JNICALL
 Java_nsk_jvmti_GetObjectMonitorUsage_objmonusage001_check(JNIEnv *env,
```
The parameters of the `check()` method are named short. It is why I decided to skip this update.
Now, I decided to rename parameters to keep them consistent with the other test `objmonusage003`.
So, it has been fixed now, thank you for the suggestion.

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1515,7 +1515,7 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
-    if (ret.waiter_count > 0) {
+    if (ret.waiter_count > 0 || ret.notify_waiter_count > 0) {
       // we have contending and/or waiting threads
       if (nWant > 0) {
         // we have contending threads
```
Yes, I've figured this out now. Thank you for pointing to it.
It feels, the counts can be calculated correctly without touching the implementation of `current_pending_monitor()`, `current_waiting_monitor()`, `_contensions` and `_waiters`.
At least, I'll try to fix it locally in the `JvmtiEnvBase::get_object_monitor_usage()`.
Please, let me know what do you prefer.

dcubed-ojdk replied:
I don't have a preference (yet). Like what David suggested, I think we need to
determine the list of thread and monitor interaction scenarios that want to
exercise with this API. Once we're happy that those scenarios represent the
complete list of possible combinations we should double check that against the
API spec to make sure that those scenarios cover the API spec. Finally, we need
to make sure that we have a test that covers all those scenarios.

It has been a long, long time since I've looked at those NSK tests... :-(

sspitsyn replied:
Thank you, Dan!
I have a fix and will push it after the mach5 tiers testing.
I've extended the test `test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003` to have all combinations that David suggested but it was a little bit tricky to make it stable. It feels that an update of the current_pending_monitor() to report the re-entered monitors would make it more elegant. But I'm not sure how to do it correctly yet. Also, I've noticed we do not post the JVMTI events  `MonitorContendedEnter` and `MonitorContendedEntered` for the re-enter case. It looks like a bug to me.

sspitsyn commented:
> I think the only way to make sense of this is to actually set up scenarios where we have different threads contending on entry, different threads waiting and different threads re-entering after being notified, and see what values actually get returned in each case.

I've added this kind of test coverage to the test NSK test:
  `test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003`
 
 > I think the `pending_current_monitor` issue may be a distinct/different issue.
 
 I tried to find a solution fix the `pending_current_monitor` to cover the monitor re-enter case and found that it is not clear how to do it in a straight-forward way. So, decide to leave it alone for now. However, it seems, it could be more elegant to fix this function. I still can make it a try to do that if you give me some hints or ideas on how to do it better.

sspitsyn commented:
I've pushed an update addressing review comments from Dan/David.
This is a short list describing the updates:
 - now the function `Threads::get_pending_threads()` in the `threads.cpp` is supporting the monitor re-enter case
 - introduced the function `oop JavaThread::vthread_or_thread()` used in the function `Threads::get_pending_threads()`
 - the fix in `jvmtiEnvBase.cpp` has been re-arranged to calculate early the right sizes of the `waiters` and `notify_waiters` array
 - the NSK test `test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003` was extended to provide coverage suggested by David. It recreates a scenario with several threads blocked on the monitor enter, several threads blocked on re-enter and several thread waiting in the `Object.wait()` to be notified.
 - some small test fixes suggested in review

dholmes-ora commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003.java:
```diff
@@ -28,9 +28,9 @@
-    final static int NUMBER_OF_THREADS = 16;
-    final static int WAIT_TIME = 100;
-    volatile static boolean waiterInLockCheck = false;
+    final static int NUMBER_OF_ENTERER_THREADS = 4;
```
Nit: ENTERING not ENTERER

sspitsyn replied:
Okay, thanks. Will rename it.

dholmes-ora commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003.java:
```diff
@@ -28,9 +28,9 @@
-    final static int WAIT_TIME = 100;
-    volatile static boolean waiterInLockCheck = false;
+    final static int NUMBER_OF_ENTERER_THREADS = 4;
+    final static int NUMBER_OF_WAITER_THREADS  = 4;
+    final static int NUMBER_OF_THREADS = NUMBER_OF_ENTERER_THREADS + NUMBER_OF_WAITER_THREADS;
```
This testing is not sufficient. We have three states of interest:
- entering the monitor directly
- waiting in the monitor via Object.wait
- re-rentering the monitor after being notified (or timed-out or interrupted)
- we need to check all of these conditions in permutations covering zero and non-zero for each one, and then see that we get the correct counts back from JVM TI.

sspitsyn replied:
The last suggestion has been addressed by adding missing test cases to the `nsk/jvmti/GetObjectMonitorUsage/objmonusage003` test.

dholmes-ora reviewed (commented):
Sorry really struggling to understand this now. We have gone from a simple miscalculation to apparently doing everything wrong.

IIUC this API does not currently handle virtual threads correctly -i s that the case? If so I would like to see that fix factored out and done separately before this fix is done. Thanks.

sspitsyn commented on src/hotspot/share/runtime/javaThread.cpp:
```diff
@@ -189,6 +189,16 @@ void JavaThread::set_jvmti_vthread(oop p) {
+// If there is a virtual thread mounted then return vthread() oop.
+// Otherwise, return threadObj().
+oop JavaThread::vthread_or_thread() const {
+    oop result = vthread();
+    if (result == nullptr) {
```
Thank you for the question.
This function does not care if this is a vthread, or not and if it is mounted or not.
We just need a right oop of currently executed thread or vthread.
The `threadOop()` and `vthread()` can return the same oop.
If `vthread()` is not `nullptr` then it can point either to a `j.l.Thread` or a `j.lVirtualThread`.
If it is `nullptr` then we have to take `threadOop()`.

sspitsyn commented:
> Sorry really struggling to understand this now. We have gone from a simple miscalculation to apparently doing everything wrong. IIUC this API does not currently handle virtual threads correctly - is that the case? If so I would like to see that fix factored out and done separately before this fix is done. Thanks.

Thank you for the concern, David. But it is not clear what it is. Could you, please, explain a little bit?
Why do you think the virtual threads are handled incorrectly?

dcubed-ojdk commented:
@sspitsyn - When you get the chance, can you checkout these possible
changes for the objmonusage001 test?

https://github.com/openjdk/jdk/pull/17839

Summary of my test changes:
- extend the check() function to verify more of the jvmtiMonitorUsage fields
- add detailed comments at each of the verification points so it's clear what is being verified and why
- add a better summary for the overall test
- make it clear which output lines document failures

I've also made some temporary debugging changes that are only useful
while updating this test:
- temporarily use the "printdump" parameter to enable test output
- add a flag to verify the current JVMs bad jvmtiMonitorUsage fields so that the test can pass with an unmodified JVM
- add a flag to add a delay that helps check for a lack of races in the verification points

These proposed test changes help document how I think the GetObjectMonitorUsage API
is supposed to work on a simple test case.

dholmes-ora commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -1194,7 +1194,11 @@ GrowableArray<JavaThread*>* Threads::get_pending_threads(ThreadsList * t_list,
     address pending = (address)p->current_pending_monitor();
-    if (pending == monitor) {             // found a match
+    address waiting = (address)p->current_waiting_monitor();
+    if (pending == monitor ||
+        (waiting == monitor && JavaThreadStatus::BLOCKED_ON_MONITOR_ENTER ==
+         java_lang_Thread::get_thread_status(p->vthread_or_thread()))
```
This code seems extremely racy. Is there anything that is stopping the target thread from running while we attempt this inspection of the oop state? If it was waiting and was interrupted or timed-out then we could mistakenly think it was still pending entry to this monitor when in fact it is not.
It is also a shame that we have to reach up into the oop thread state to figure out if if it is blocked on re-entry ...

sspitsyn replied:
The function `get_pending_threads()` is executed only by the VM_GetObjectMonitorUsage operation.
I've added this assert there: 
    `assert(Thread::current()->is_VM_thread(), "Must be the VM thread");`
Does it address your concern?

dholmes-ora replied:
Yes thanks. Didn't realize we were in a safepoint VM op.

dholmes-ora reviewed (commented):
> Why do you think the virtual threads are handled incorrectly?

Sorry my mistake. The `thread_or_vthread` changes had me confused.

sspitsyn commented:
> When you get the chance, can you checkout these possible
changes for the objmonusage001 test?

Thanks, Dan. I've pushed the suggested test changes but refactored them a little bit.

sspitsyn commented:
From @dholmes-ora :
> If you don't check all the zero/non-zero permutations for the three counts of interest then you don't know that you are combining them the right way to give the two counts reported by the API.

Okay, thanks. I've added the suggested test cases to the `nsk/jvmti/GetObjectMonitorUsage/objmonusage003`.
One of the testcases is covered by the `nsk/jvmti/GetObjectMonitorUsage/objmonusage001`.

dcubed-ojdk commented:
> Thanks, Dan. I've pushed the suggested test changes but refactored them a little bit.

You are welcome. I presume the revised test passes with your fix in place. I made the
test changes on a baseline repo and not a repo that had your changes which is why I
had temporary debug flags in the test..

Also, do you agree with my test assertion comments in the verification point comments?
Am I properly understanding how we are changing this API?

sspitsyn commented:
From @dcubed-ojdk : 
> Also, do you agree with my test assertion comments in the verification point comments?
> Am I properly understanding how we are changing this API?

Yes, I agree. All looks right to me.
Q: Do you want to keep the `CHECK_FOR_BAD_RESULTS` mode or it can be removed after the failing output was verified?

dcubed-ojdk commented:
CHECK_FOR_BAD_RESULTS and ADD_DELAYS_FOR_RACES are for
temporary debugging only.

sspitsyn commented:
I've pushed an update that includes:
 - added minor spec clarification for JDWP `ObjectReference.monitor_info` command (please, see the CSR)
 - removed the CHECK_FOR_BAD_RESULTS debugging mode from the test: `nsk/jvmti/GetObjectMonitorUsage/objmonusage001`

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1485,14 +1485,28 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
   }
 
+  ResourceMark rm(current_thread);
+  GrowableArray<JavaThread*>* wantList = Threads::get_pending_threads(tlh.list(), nWant + nWait, (address)mon);
+  nWant = wantList->length();
+
+  if (mon != nullptr) {
```
Shouldn't the call to `get_pending_threads` only happen if `mon != nullptr`? Otherwise the `wantList` has to be empty.

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1485,14 +1485,28 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
+    ObjectWaiter *waiter = mon->first_waiter();
+    for (int i = 0; i < nWait; i++) {
+      if (waiter == nullptr || (i != 0 && waiter == mon->first_waiter())) {
+        // robustness: the waiting list has gotten smaller
```
We are at a safepoint so I don't see how the wait list can shrink. I initially thought perhaps a waiter could timeout, but the code that does the timed park is wrapped in ` ThreadBlockInVMPreprocess` which will block at a safepoint if one is active.

sspitsyn replied:
Thank you for the question.
The `nWait` count we got from the `mon->waiters()` can include the threads that are re-entering the monitor after being notified. Here we are correcting the actual number of the waiting threads by excluding those re-entering the monitor. The comment need an update to explain it.

dholmes-ora reviewed (commented):
Thanks for the test updates.

A couple of other queries.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,249 @@
+    final static int NUMBER_OF_THREADS = NUMBER_OF_ENTERING_THREADS + NUMBER_OF_WAITING_THREADS;
+
+    static Object lockCheck = new Object();
+    static Thread thr[] = new Thread[NUMBER_OF_THREADS];
```
Nit: the `[]` go on the type not the variable.

sspitsyn replied:
Thanks, fixed now.

sspitsyn commented:
Leonid, thank you for the comments. Will work on addressing them.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,249 @@
+
+public class ObjectMonitorUsage {
+
+    final static int JCK_STATUS_BASE = 95;
```
Thanks, done.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,162 @@
+    if (inf.waiter_count > 0) {
+        LOG(">>>       waiters:\n");
+        for (j = 0; j < inf.waiter_count; j++) {
+          err = jvmti->GetThreadInfo(inf.waiters[j], &tinf);
```
Done, thanks.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,249 @@
+     * - zero re-entering threads
+     * - zero threads waiting to be notified
+     */
+    static void test1(boolean isVirtual) throws Error {
```
Fixed, thanks.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,162 @@
+    if (inf.owner == nullptr) {
+      LOG(">>> [%2d]    owner: none (0x0)\n", count);
+    } else {
+      err = jvmti->GetThreadInfo(inf.owner, &tinf);
```
Fixed, thanks.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,162 @@
+  if (err == JVMTI_ERROR_MUST_POSSESS_CAPABILITY && !caps.can_get_monitor_info) {
+    return; /* Ok, it's expected */
+  } else if (err != JVMTI_ERROR_NONE) {
+    LOG("(GetMonitorInfo#%d) unexpected error: %s (%d)\n",
```
Fixed, thanks.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,162 @@
+
+  err = jvmti->GetObjectMonitorUsage(obj, &inf);
+  if (err == JVMTI_ERROR_MUST_POSSESS_CAPABILITY && !caps.can_get_monitor_info) {
+    return; /* Ok, it's expected */
```
Yes. Fixed, thanks.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,249 @@
+        test3(false);
+
+        // test virtual threads
+        test1(false);
```
Nice catch, thanks. Fixed now.

sspitsyn commented on test/hotspot/jtreg/serviceability/jvmti/GetObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,249 @@
+ *       - unowned object with waitings through Object.wait()
+ *       - all the above is checked for both platform and virtual threads
+ * @requires vm.jvmti
+ * @compile ObjectMonitorUsage.java
```
Fixed, thanks.

sspitsyn commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003.java:
```diff
@@ -28,9 +28,9 @@
 public class objmonusage003 {
 
     final static int JCK_STATUS_BASE = 95;
-    final static int NUMBER_OF_THREADS = 16;
```
Agreed. I was thinking about the same.

sspitsyn commented:
I've pushed an update which addresses test related review comments from Leonid and adds some test improvements:
 - remove the `nsk/jvmti` test `objmonusage003` which was converted into ObjectMonitorUsage
 - rename the test folder from `GetObjectMonitorUsage` to `ObjectMonitorUsage` for naming unification
 - add one more testing scenario: `test0`
 - add test specific thread names
 - address several minor review comments from Leonid
 - add better organized log messages

lmesnik commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -1181,10 +1182,12 @@ void Threads::metadata_handles_do(void f(Metadata*)) {
 }
 
-// Get count Java threads that are waiting to enter the specified monitor.
+#if INCLUDE_JVMTI
```
Is it needed to update thread.hpp also?

sspitsyn replied:
Thank you for the comment, will check it.

lmesnik reviewed (approved):
There is one comment about changing #define for thread.hpp.
The test changes look good for me. Thanks for converting the test from vmTetbase.

sspitsyn commented:
Thank you for review, Leonid!

sspitsyn commented:
@dcubed-ojdk @dholmes-ora
Dan and David, I think, I've addressed all your comments.
Do you have any other concerns? Can you approve it now?

dholmes-ora commented:
> The notify_waiters has to include all threads waiting to be notified in Object.wait().
> The implementation also includes the threads waiting to re-enter the monitor in Object.wait() which is wrong.

This is not in fact the case so can you please remove this from the problem statement/description above.

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1479,20 +1479,38 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
     assert(mon != nullptr, "must have monitor");
     // this object has a heavyweight monitor
     nWant = mon->contentions(); // # of threads contending for monitor
     nWait = mon->waiters();     // # of threads in Object.wait()
```
Please clarify the comment as
```
// # of threads waiting for notification, or to re-enter monitor, in Object.wait()
```

sspitsyn replied:
Good catch, thanks. Will fix.

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1479,20 +1479,38 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
     mon = mark.monitor();
     assert(mon != nullptr, "must have monitor");
     // this object has a heavyweight monitor
     nWant = mon->contentions(); // # of threads contending for monitor
```
Please clarity comment as
```
// # of threads contending for monitor entry, but not -rentry
```

sspitsyn replied:
Will fix, thanks.

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1479,20 +1479,38 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
     nWait = mon->waiters();     // # of threads in Object.wait()
-    ret.waiter_count = nWant + nWait;
-    ret.notify_waiter_count = nWait;
+    wantList =  Threads::get_pending_threads(tlh.list(), nWant + nWait, (address)mon);
```
Please add a comment
```
// Get the actual set of threads trying to enter, or re-enter, the monitor.
```

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1479,20 +1479,38 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
+    for (int i = 0; i < nWait; i++) {
+      if (waiter == nullptr || (i != 0 && waiter == mon->first_waiter())) {
+        // robustness: the waiting list has gotten smaller
+        // The nWait count we got from the mon->waiters() may include the re-entering
+        // the monitor threads after being notified. Here we are correcting the actual
+        // number of the waiting threads by excluding those re-entering the monitor.
+        nWait = i;
```
Sorry I can't follow the logic here. Why is the whole block not simply:
```
if (mon != nullptry) {
  // The nWait count may include threads trying to re-enter the monitor, so
  // walk the actual set of waiters to count those actually waiting.
  int count = 0;
  for (ObjectMonitor * waiter = mon->first_waiter(); waiter != nullptr; waiter = mon->next_waiter(waiter)) {
    count++;
  }
}
ret.notify_waiter_count = count;
```

sspitsyn replied:
Thank you for the suggestion. Let me check this.

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1510,56 +1528,33 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
+      for (int i = 0; i < nWait; i++) {
+        JavaThread *w = mon->thread_of_waiter(waiter);
+        assert(w != nullptr, "DBG: sanity check");
+        // If the thread was found on the ObjectWaiter list, then
+        // it has not been notified. This thread can't change the
+        // state of the monitor so it doesn't need to be suspended.
```
Not sure why suspension is being mentioned here.

sspitsyn replied:
Removed this statement from the comment. Will restore in a case of Dan's objection.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,326 @@
+ * @summary converted from VM Testbase nsk/jvmti/GetObjectMonitorUsage/objmonusage003
+ * DESCRIPTION
+ *     The test checks if the JVMTI function GetObjectMonitorUsage returns
+ *     the expected values for the owner, entry_count, water_count
```
s/water_count/waiter_count/

sspitsyn replied:
Thanks, fixed now.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,326 @@
+ *     the expected values for the owner, entry_count, water_count
+ *     fields of JVMTI_monitor_info.
+ *     The testcases are the following:
+ *       - unowned object without any waitings
+ *       - unowned object with waitings to be notified
+ *       - owned object without any waitings
+ *       - owned object with N waitings to enter the monitor
+ *       - owned object with N waitings to be notified
+ *       - owned object with N waitings to enter, from 0 to N waitings to re-enter,
+ *         from N to 0 waitings to be notified
```
"waitings" is not grammatically valid. Suggestion: s/waitings/threads waiting/

sspitsyn replied:
Thanks, fixed now.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,326 @@
+ *       - owned object with N waitings to enter, from 0 to N waitings to re-enter,
+ *         from N to 0 waitings to be notified
+ *       - all the above scenarios are executed with platform and virtual threads
+ * @requires vm.continuations
```
Do we really require this?

sspitsyn replied:
Yes, can be removed. Thanks.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,326 @@
+            synchronized (lockCheck) {
+                try {
+                    ready = true;
+                    lockCheck.wait();
```
Spurious wakeups will break this test. They shouldn't happen but ...

sspitsyn replied:
Thanks, will check how this can be fixed.

dholmes-ora replied:
Sorry I missed this response. I can't see a way to address spurious wakeups in this case as it needs to be a per-thread flag (so that each thread knows it was notified) but you don't know which thread will be notified in any given call to `notify()`. I also can't see how you can detect a spurious wakeup in this code. If they happen then a subtest may fail due to an unexpected number of re-entering threads.
I think we will just have to see how stable the test is in practice.

sspitsyn replied:
Okay, thanks! I'll add some diagnostic code to catch spurious wakups.
Let's see if we ever encounter any spurious wakeup in this test.

dcubed-ojdk replied:
I think some sort of comment needs to be added here to document
the possibility of this code path being affect by a spurious wakeup.

sspitsyn replied:
Thanks. I've added both a comment and also some spurious wakeup detection with a warning.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -0,0 +1,326 @@
+        String vtag = vtag(isVirtual);
+        log("\n###test0: started " + vtag);
+
+        Thread[] wThreads = startWaitingThreads(isVirtual);
+
+        // entry count: 0
+        // count of threads waiting to enter:       0
+        // count of threads waiting to re-enter:    0
+        // count of threads waiting to be notified: NUMBER_OF_WAITING_THREADS
+        check(lockCheck, null, 0, // no owner thread
```
AFAICS you are racing here, as the waiting threads can set `ready` but not yet have called `wait()`. To know for sure they are in `wait()` you need to acquire the monitor before checking (and release it again so that you test the no-owner scenario as intended). But this should probably be done inside `startWaitingThreads`.

sspitsyn replied:
Thank you for the comment. Will implement similar approach as in the `startEnteringThreads()`.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,219 @@
+}
+
+
+jint  Agent_Initialize(JavaVM *jvm, char *options, void *reserved) {
```
Nit: there's a double space after jint

sspitsyn replied:
Will fix, thanks,

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,219 @@
+  return result;
+}
+
+} // exnern "C"
```
typo: exnern -> extern

sspitsyn replied:
Will fix, thanks.

dholmes-ora reviewed (commented):
Thanks Serguei, this is generally looking good, but I have one main query with the logic, a number of suggestions regarding comments, and a few issues in the new test code.

dholmes-ora commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1486,26 +1486,27 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
+    // number of the waiting threads by excluding those re-entering the monitor.
+    nWait = 0;
+    for (ObjectWaiter* waiter = mon->first_waiter();
+         waiter != nullptr && (nWait == 0 || waiter != mon->first_waiter());
```
Sorry I do not understand the logic of this line. `waiters` is just a linked-list of `ObjectWaiter`s so all we need to do is traverse it and count the number of elements.

sspitsyn replied:
The loop is endless without this extra condition, so we are getting a test execution timeout.
The `waiters` seems to be `circular doubly linked list` as we can see below:
```
inline void ObjectMonitor::AddWaiter(ObjectWaiter* node) {
  assert(node != nullptr, "should not add null node");
  assert(node->_prev == nullptr, "node already in list");
  assert(node->_next == nullptr, "node already in list");
  // put node at end of queue (circular doubly linked list)
  if (_WaitSet == nullptr) {
    _WaitSet = node;
    node->_prev = node;
    node->_next = node;
  } else {
    ObjectWaiter* head = _WaitSet;
    ObjectWaiter* tail = head->_prev;
    assert(tail->_next == head, "invariant check");
    tail->_next = node;
    head->_prev = node;
    node->_next = head;
    node->_prev = tail;
  }
}
```
I'll make sure nothing is missed and check the old code again.

sspitsyn replied:
Okay, thanks.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -28,18 +28,17 @@
- *       - owned object with N waitings to be notified
- *       - owned object with N waitings to enter, from 0 to N waitings to re-enter,
- *         from N to 0 waitings to be notified
+ *       - unowned object without any waiting threads
+ *       - unowned object with threads waiting to be notified
+ *       - owned object without any waiting threads
```
You could say "threads waiting" in all cases - it looks odd to reverse it for two cases.

sspitsyn replied:
Thanks, updated.

dholmes-ora reviewed (commented):
Updates look good. A couple of minor commenst.

dholmes-ora reviewed (approved):
Looks good. Thanks.

Sorry it has taken me so long to work through this one. It seemed so simple when I filed the bug :)

sspitsyn commented:
> Sorry I missed this response. I can't see a way to address spurious wakeups in this case as it needs to be a per-thread flag (so that each thread knows it was notified) but you don't know which thread will be notified in any given call to notify(). I also can't see how you can detect a spurious wakeup in this code. If they happen then a subtest may fail due to an unexpected number of re-entering threads. I think we will just have to see how stable the test is in practice.

Okay, thanks! Let's see if we ever encounter any spurious wakeup in this test.

dcubed-ojdk commented:
I last reviewed v02 of this PR so I've jumped forward to the full v22 webrev. It may take
me a bit of time to redo this review.

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1479,20 +1479,39 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
 
   jint nWant = 0, nWait = 0;
   markWord mark = hobj->mark();
+  ResourceMark rm(current_thread);
+  GrowableArray<JavaThread*>* wantList = nullptr;
```
Thanks for refactoring `JvmtiEnvBase::get_object_monitor_usage()` to
be much more clear. It was difficult to review in GitHub so I switched to
the webrev frames view so that I could scroll the two sides back and forth
and determine equivalence.

sspitsyn replied:
Good!

dcubed-ojdk commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -1181,10 +1182,12 @@ void Threads::metadata_handles_do(void f(Metadata*)) {
 
-// Get count Java threads that are waiting to enter the specified monitor.
+#if INCLUDE_JVMTI
+// Get count Java threads that are waiting to enter or re-enter the specified monitor.
```
nit typo: s/count Java/count of Java/

sspitsyn replied:
Thanks, fixed now.

dcubed-ojdk commented on src/hotspot/share/runtime/threads.hpp:
```diff
@@ -131,7 +131,7 @@ class Threads: AllStatic {
   static unsigned print_threads_compiling(outputStream* st, char* buf, int buflen, bool short_form = false);
 
-  // Get Java threads that are waiting to enter a monitor.
+  // Get count Java threads that are waiting to enter or re-enter the specified monitor.
```
nit typo: s/count Java/count of Java/

Also this file needs a copyright year update.

sspitsyn replied:
Thanks, fixed now.

dcubed-ojdk commented on test/hotspot/jtreg/TEST.quick-groups:
```diff
@@ -969,7 +969,6 @@ vmTestbase_nsk_jvmti_quick = \
   vmTestbase/nsk/jvmti/GetMethodName/methname003/TestDescription.java \
   vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage001/TestDescription.java \
   vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage002/TestDescription.java \
-  vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage003/TestDescription.java \
```
This file needs a copyright year update.

sspitsyn replied:
Thanks, fixed now.

dcubed-ojdk commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetObjectMonitorUsage/objmonusage001/TestDescription.java:
```diff
@@ -29,13 +29,20 @@
  * @library /vmTestbase
  *          /test/lib
- * @run main/othervm/native -agentlib:objmonusage001 nsk.jvmti.GetObjectMonitorUsage.objmonusage001
+ * @run main/othervm/native -agentlib:objmonusage001=printdump nsk.jvmti.GetObjectMonitorUsage.objmonusage001
```
I only included this `=printdump` for debug when I was updating the test.
You don't have to keep it.

sspitsyn replied:
Thanks, removed now.

dcubed-ojdk commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,249 @@
+
+  print_monitor_info(jni, inf);
+
+  if (!jni->IsSameObject(owner, inf.owner)) {
+    LOG("(%d) unexpected owner: 0x%p\n", check_idx, inf.owner);
+    result = STATUS_FAILED;
+  }
+  if (inf.entry_count != entryCount) {
+    LOG("(%d) entry_count expected: %d, actually: %d\n",
+        check_idx, entryCount, inf.entry_count);
+    result = STATUS_FAILED;
+  }
+  if (inf.waiter_count != waiterCount) {
+    LOG("(%d) waiter_count expected: %d, actually: %d\n",
+        check_idx, waiterCount, inf.waiter_count);
+    result = STATUS_FAILED;
+  }
+  if (inf.notify_waiter_count != notifyWaiterCount) {
+    LOG("(%d) notify_waiter_count expected: %d, actually: %d\n",
+        check_idx, notifyWaiterCount, inf.notify_waiter_count);
+    result = STATUS_FAILED;
```
Please consider prefixing these LOG output lines with "FAILED:" so that it
is easier to identify the log output where failure info is shown.

sspitsyn replied:
Good suggestion. Fixed now.

dcubed-ojdk commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,249 @@
+
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+
+  if (tested_monitor != nullptr) {
+    jni->DeleteGlobalRef(tested_monitor);
+  }
```
Since this is at the beginning of `setTestedMonitor` does it mean that we "leak"
the last global ref?

sspitsyn replied:
Thank you for the check. There is no leak of the last global ref because there is always this call at the end of sub-test: `setTestedMonitor(null)`.

dcubed-ojdk commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp:
```diff
@@ -0,0 +1,249 @@
+  return jni->IsSameObject(monitor, tested_monitor) == JNI_TRUE;
+}
+
+JNIEXPORT void JNICALL
+MonitorContendedEnter(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_enter++;
+  }
+}
+
+JNIEXPORT void JNICALL
+MonitorContendedEntered(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_enter--;
+  }
+}
+
+JNIEXPORT void JNICALL
+MonitorWait(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor, jlong timeout) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_be_notified++;
+  }
+}
+
+JNIEXPORT void JNICALL
+MonitorWaited(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jobject monitor, jboolean timed_out) {
+  RawMonitorLocker rml(jvmti, jni, event_lock);
+  if (is_tested_monitor(jni, monitor)) {
+    waits_to_be_notified--;
+  }
+}
```
I like these for getting rid of the raciness of knowing when a thread is blocked
on contention, finished entering, blocked on wait or finished waiting. Nice!

dcubed-ojdk reviewed (approved):
Thumbs up on the v22 version of this change. Of course, I'm assuming
that the updated tests pass and that there are no new failures in the
usual Mach5 testing.

Sorry for the delay in getting back to this review. A combination of
vacation following by illness delayed me getting back to this PR.

Please note that I did not try to compare the (deleted) objmonusage003
version of the test with the (new) ObjectMonitorUsage version. I didn't
think that would be useful.

sspitsyn commented:
Thank you for review, Dan! I believe, all you comments have been addressed now.

dholmes-ora commented on test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:
```diff
@@ -322,6 +345,13 @@ public void run() {
                     lockCheck.wait();
+
+                    // identify if the wakeup was spurious
+                    wokeupCount++;
+                    if (!notifiedAll && notifiedCount < wokeupCount) {
+                        wasSpuriousWakeup = true;
```
This doesn't work. First as you've realized you cannot detect a spurious wakeup when notifyAll is used. But second the `notifiedCount` is incremented on each `notify` whilst the monitor lock is still held - as it must be. So no `wait` can return until the monitor is unlocked. So unless the spurious wakeup occurs before the monitor is acquired at line 254, all notifications must be completed before even a spurious wakeup can cause `wait()` to return - at which point `notifiedCount` must equal ` NUMBER_OF_WAITING_THREADS` which can never be < `wokeupCount`.
Given you can only detect one very specific case of a spurious wakeup, all this extra logic is just adding to the complexity and giving a false impression about what is being checked. To detect actual spurious wakeups here you need to be able to track when each thread was chosen by a `notify()` and there is no means to do that.
